### PR TITLE
remove EmptyObject usage

### PR DESCRIPTION
**Pull request by ghost**
### Fixed Issues

$ https://github.com/Expensify/App/issues/39124
PROPOSAL: https://github.com/Expensify/App/issues/39124#issuecomment-2062102119

### Tests

- [x] Verify that no errors appear in the JS console

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Review by rayane-d — CHANGES_REQUESTED** [review]
Parial review

**Review by rayane-d — CHANGES_REQUESTED** [review]
Lint and TypeScript Checks are failing

**Comment by blazejkustra** [comment]
Lint and TS checks are failing, could you have a look @jsdev2547?

**Comment by jsdev2547** [comment]
@rayane-djouah @blazejkustra I have resolved lint and typescript errors. PR is ready for review.

File: src/libs/actions/IOU.ts [hunk 1]
```diff
@@ -1968,7 +1967,7 @@ function getTrackExpenseInformation(
     moneyRequestReportID = '',
     linkedTrackedExpenseReportAction?: OnyxTypes.ReportAction,
     existingTransactionID?: string,
-): TrackExpenseInformation | EmptyObject {
+): OnyxEntry<TrackExpenseInformation> {
```
Comment (fabioh8010) [inline src/libs/actions/IOU.ts:1970]:
```suggestion
): TrackExpenseInformation | null {
```

`TrackExpenseInformation` is not a value stored in Onyx

File: src/libs/actions/IOU.ts [hunk 2]
```diff
@@ -2117,7 +2116,7 @@ function getTrackExpenseInformation(
         iouAction,
         reportPreviewAction,
         optimisticTransactionThread ?? {},
-        (optimisticCreatedActionForTransactionThread as OptimisticCreatedReportAction) ?? {}, // Add type assertion here
+        optimisticCreatedActionForTransactionThread ?? null, // Add type assertion here
```
Comment (fabioh8010) [inline src/libs/actions/IOU.ts:2119]:
```suggestion
        optimisticCreatedActionForTransactionThread ?? null,
```

Let's remove this comment

File: src/libs/actions/IOU.ts [hunk 3]
```diff
@@ -3094,10 +3095,10 @@ function categorizeTrackedExpense(
     taxCode = '',
     taxAmount = 0,
     billable?: boolean,
-    receipt?: Receipt,
+    receipt: OnyxEntry<Receipt> = null,
```
Comment (fabioh8010) [inline src/libs/actions/IOU.ts:3098]:
```suggestion
    receipt: Receipt | null = null,
```

`Receipt` is not a value stored in Onyx

File: src/libs/actions/IOU.ts [hunk 4]
```diff
@@ -3171,10 +3172,10 @@ function shareTrackedExpense(
...
```
Comment (fabioh8010) [inline src/libs/actions/IOU.ts:3175]:
```suggestion
    receipt: Receipt | null = null,
```

File: src/libs/actions/IOU.ts [hunk 5]
```diff
@@ -3371,7 +3372,7 @@ function sendInvoice(
     currentUserAccountID: number,
     transaction: OnyxEntry<OnyxTypes.Transaction>,
     invoiceChatReport?: OnyxEntry<OnyxTypes.Report>,
-    receiptFile?: Receipt,
+    receiptFile?: OnyxEntry<Receipt>,
```
Comment (fabioh8010) [inline src/libs/actions/IOU.ts:3375]:
```suggestion
    receiptFile?: Receipt | null,
```

File: src/libs/actions/IOU.ts [hunk 6]
```diff
@@ -3426,7 +3427,7 @@ function trackExpense(
     payeeAccountID: number,
     participant: Participant,
     comment: string,
-    receipt?: Receipt,
+    receipt: OnyxEntry<Receipt> = null,
```
Comment (fabioh8010) [inline src/libs/actions/IOU.ts:3430]:
```suggestion
    receipt: Receipt | null = null,
```

File: src/libs/API/parameters/CategorizeTrackedExpenseParams.ts [hunk 7]
```diff
@@ -16,7 +17,7 @@ type CategorizeTrackedExpenseParams = {
     reportPreviewReportActionID: string;
     category?: string;
     tag?: string;
-    receipt?: Receipt;
+    receipt: OnyxEntry<Receipt>;
```
Comment (fabioh8010) [inline src/libs/API/parameters/CategorizeTrackedExpenseParams.ts:20]:
```suggestion
    receipt: Receipt | null;
```

`Receipt` is not a value stored in Onyx

File: src/libs/API/parameters/ShareTrackedExpenseParams.ts [hunk 8]
```diff
@@ -16,7 +17,7 @@ type ShareTrackedExpenseParams = {
     reportPreviewReportActionID: string;
     category?: string;
     tag?: string;
-    receipt?: Receipt;
+    receipt: OnyxEntry<Receipt>;
```
Comment (fabioh8010) [inline src/libs/API/parameters/ShareTrackedExpenseParams.ts:20]: Same

File: src/libs/API/parameters/TrackExpenseParams.ts [hunk 9]
```diff
@@ -15,7 +16,7 @@ type TrackExpenseParams = {
     createdChatReportActionID: string;
     createdIOUReportActionID?: string;
     reportPreviewReportActionID?: string;
-    receipt?: Receipt;
+    receipt?: OnyxEntry<Receipt>;
```
Comment (fabioh8010) [inline src/libs/API/parameters/TrackExpenseParams.ts:19]: Same

**Review by fabioh8010 — COMMENTED** [review]
@jsdev2547 Left some comments, let's address @rayane-djouah ones too.

Also, I've noticed in the codebase several usages of `Record<string, never>`, which is basically the `EmptyObject` we are dealing in this PR.

Let's remove `Record<string, never>` usage in the following files:

- src/components/withCurrentUserPersonalDetails.tsx
- src/hooks/useCurrentUserPersonalDetails.ts
- src/libs/ModifiedExpenseMessage.ts
- src/libs/ReportActionsUtils.ts
- src/libs/ReportUtils.ts
- src/libs/actions/Task.ts
- src/pages/home/report/withReportAndReportActionOrNotFound.tsx

And, replace `Record<string, never>` with `EmptyObject` in the following ones:

- src/components/PopoverProvider/types.ts
- src/hooks/useKeyboardShortcut.ts
- src/libs/migrations/CheckForPreviousReportActionID.ts
- src/types/onyx/OriginalMessage.ts

**Comment by fabioh8010** [comment]
Also @jsdev2547 please resolve conflicts!

**Comment by jsdev2547** [comment]
@rayane-djouah @fabioh8010 I have resolved the conflicts and also resolved the comments. PR is ready review.

File: src/hooks/useKeyboardShortcut.ts [hunk 10]
```diff
@@ -26,7 +27,7 @@ type KeyboardShortcutConfig = {
  * Register a keyboard shortcut handler.
  * Recommendation: To ensure stability, wrap the `callback` function with the useCallback hook before using it with this hook.
  */
-export default function useKeyboardShortcut(shortcut: Shortcut, callback: (e?: GestureResponderEvent | KeyboardEvent) => void, config: KeyboardShortcutConfig | Record<string, never> = {}) {
+export default function useKeyboardShortcut(shortcut: Shortcut, callback: (e?: GestureResponderEvent | KeyboardEvent) => void, config: KeyboardShortcutConfig | EmptyObject = {}) {
```
Comment (blazejkustra) [inline src/hooks/useKeyboardShortcut.ts:30]:
```suggestion
export default function useKeyboardShortcut(shortcut: Shortcut, callback: (e?: GestureResponderEvent | KeyboardEvent) => void, config: KeyboardShortcutConfig = {}) {
```

File: src/components/PopoverProvider/types.ts [hunk 11]
```diff
@@ -1,14 +1,15 @@
...
+import type {EmptyObject} from '@src/types/utils/EmptyObject';
 
 type PopoverContextProps = {
     children: ReactNode;
 };
 
 type PopoverContextValue = {
     onOpen?: (popoverParams: AnchorRef) => void;
-    popover?: AnchorRef | Record<string, never> | null;
+    popover?: AnchorRef | EmptyObject | null;
```
Comment (blazejkustra) [inline src/components/PopoverProvider/types.ts:12]:
```suggestion
    popover?: AnchorRef | null;
```

File: src/components/LHNOptionsList/OptionRowLHNData.tsx [hunk 12]
```diff
@@ -24,7 +24,7 @@ function OptionRowLHNData({
     receiptTransactions,
     parentReportAction,
     transaction,
-    lastReportActionTransaction = {},
+    lastReportActionTransaction = null,
```
Comment (blazejkustra) [inline src/components/LHNOptionsList/OptionRowLHNData.tsx:27]: We don't need null/undefined default values (please do the same for other places in your code if applicable):
```suggestion
    lastReportActionTransaction,
```

**Review by blazejkustra — CHANGES_REQUESTED** [review]
Some NABs, other than that looks good! 🚀

File: src/components/withCurrentUserPersonalDetails.tsx [hunk 13]
```diff
@@ -4,7 +4,7 @@ import useCurrentUserPersonalDetails from '@hooks/useCurrentUserPersonalDetails'
 import getComponentDisplayName from '@libs/getComponentDisplayName';
 import type {PersonalDetails} from '@src/types/onyx';
 
-type CurrentUserPersonalDetails = PersonalDetails | Record<string, never>;
+type CurrentUserPersonalDetails = PersonalDetails;
```
Comment (fabioh8010) [inline src/components/withCurrentUserPersonalDetails.tsx:7]: Maybe let's just remove this `CurrentUserPersonalDetails` type altogether and use `PersonalDetails` instead?

File: src/hooks/useCurrentUserPersonalDetails.ts [hunk 14]
```diff
@@ -3,7 +3,7 @@ import {usePersonalDetails, useSession} from '@components/OnyxProvider';
 import CONST from '@src/CONST';
 import type {PersonalDetails} from '@src/types/onyx';
 
-type CurrentUserPersonalDetails = PersonalDetails | Record<string, never>;
+type CurrentUserPersonalDetails = PersonalDetails;
```
Comment (fabioh8010) [inline src/hooks/useCurrentUserPersonalDetails.ts:6]: Same

File: src/libs/ReportActionsUtils.ts [hunk 15]
```diff
@@ -139,15 +138,15 @@ function isReportActionSubmitted(reportAction: OnyxEntry<ReportAction>): boolean
     return reportAction?.actionName === CONST.REPORT.ACTIONS.TYPE.SUBMITTED;
 }
 
-function isModifiedExpenseAction(reportAction: OnyxEntry<ReportAction> | ReportAction | Record<string, never>): boolean {
+function isModifiedExpenseAction(reportAction: OnyxEntry<ReportAction> | ReportAction): boolean {
```
Comment (fabioh8010) [inline src/libs/ReportActionsUtils.ts:141]:
```suggestion
function isModifiedExpenseAction(reportAction: OnyxEntry<ReportAction>): boolean {
```

File: src/libs/ReportUtils.ts [hunk 16]
```diff
@@ -622,20 +621,20 @@ function isDraftReport(reportID: string | undefined): boolean {
...
  */
-function getParentReport(report: OnyxEntry<Report> | EmptyObject): OnyxEntry<Report> | EmptyObject {
-    if (!report?.parentReportID) {
-        return {};
+function getParentReport(report: OnyxEntry<Report>): OnyxEntry<Report> {
+    if (!allReports || !report?.parentReportID) {
+        return null;
     }
-    return allReports?.[`${ONYXKEYS.COLLECTION.REPORT}${report.parentReportID}`] ?? {};
+    return allReports?.[`${ONYXKEYS.COLLECTION.REPORT}${report.parentReportID}`];
```
Comment (fabioh8010) [inline src/libs/ReportUtils.ts:628]:
```suggestion
    return allReports[`${ONYXKEYS.COLLECTION.REPORT}${report.parentReportID}`];
```

File: src/libs/ReportUtils.ts [hunk 17]
```diff
@@ -823,7 +822,7 @@ function isSettled(reportID: string | undefined): boolean {
     if (!allReports || !reportID) {
         return false;
     }
-    const report: Report | EmptyObject = allReports[`${ONYXKEYS.COLLECTION.REPORT}${reportID}`] ?? {};
+    const report: OnyxEntry<Report> = allReports?.[`${ONYXKEYS.COLLECTION.REPORT}${reportID}`] ?? null;
```
Comment (fabioh8010) [inline src/libs/ReportUtils.ts:825]:
```suggestion
    const report = allReports[`${ONYXKEYS.COLLECTION.REPORT}${reportID}`];
```

File: src/libs/WorkspacesSettingsUtils.ts [hunk 18]
```diff
@@ -66,10 +66,10 @@ const getBrickRoadForPolicy = (report: Report, altReportActions?: OnyxCollection
     }
 
     // To determine if the report requires attention from the current user, we need to load the parent report action
-    let itemParentReportAction = {};
+    let itemParentReportAction;
```
Comment (fabioh8010) [inline src/libs/WorkspacesSettingsUtils.ts:69]:
```suggestion
    let itemParentReportAction: OnyxEntry<ReportAction> = null;
```

File: src/libs/actions/Task.ts [hunk 19]
```diff
@@ -803,22 +802,22 @@ function getShareDestination(reportID: string, reports: OnyxCollection<OnyxTypes
...
  */
-function getParentReport(report: OnyxEntry<OnyxTypes.Report> | EmptyObject): OnyxEntry<OnyxTypes.Report> | EmptyObject {
-    if (!report?.parentReportID) {
-        return {};
+function getParentReport(report: OnyxEntry<OnyxTypes.Report>): OnyxEntry<OnyxTypes.Report> {
+    if (!allReports || !report?.parentReportID) {
+        return null;
     }
-    return allReports?.[`${ONYXKEYS.COLLECTION.REPORT}${report.parentReportID}`] ?? {};
+    return allReports?.[`${ONYXKEYS.COLLECTION.REPORT}${report.parentReportID}`];
```
Comment (fabioh8010) [inline src/libs/actions/Task.ts:820]:
```suggestion
    return allReports[`${ONYXKEYS.COLLECTION.REPORT}${report.parentReportID}`];
```

File: src/libs/getReportPolicyID.ts [hunk 20]
```diff
@@ -14,12 +13,12 @@ Onyx.connect({
...
  */
-function getReport(reportID: string | undefined): OnyxEntry<Report> | EmptyObject {
+function getReport(reportID: string | undefined): OnyxEntry<Report> {
     if (!allReports) {
-        return {};
+        return null;
     }
 
-    return allReports?.[`${ONYXKEYS.COLLECTION.REPORT}${reportID}`] ?? {};
+    return allReports?.[`${ONYXKEYS.COLLECTION.REPORT}${reportID}`];
```
Comment (fabioh8010) [inline src/libs/getReportPolicyID.ts:21]:
```suggestion
    return allReports[`${ONYXKEYS.COLLECTION.REPORT}${reportID}`];
```

**Review by rayane-d — CHANGES_REQUESTED** [review]
@jsdev2547 Can you please address the above comments? Also, please resolve conflicts.

**Pull request closed without merging**